libs/platform/ebpf_epoch.c

```diff
@@ -10,93 +10,132 @@
 static uint32_t _ebpf_epoch_cpu_count = 0;
 
+typedef enum _ebpf_epoch_cpu_message_type
+{
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_COMPUTE_EPOCH, ///< This message contains the oldest known epoch. It is sent to each
+                                               ///< CPU, which then queries the epoch for each thread queued on that CPU
+                                               ///< and sets the epoch in the  message to the minimum of the local
+                                               ///< minima and the minima in the message. The message is then forwarded
+                                               ///< to the next CPU. The last CPU then sets _ebpf_release_epoch to the
+                                               ///< minima minus 1 and sends a re-arm timer to CPU 0.
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_REARM_TIMER, ///< This message instructs the CPU to clear the timer-armed flag, releases
+                                             ///< any items in the free list that are eligible for reclamation, re-arms
+                                             ///< the timer if need, and forwards the message to the next CPU. The last
+                                             ///< CPU then sends a epoch complete message to CPU 0.
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_COMPUTE_EPOCH_COMPLETE, ///< This message is sent to CPU 0 to indicate that computation
+                                                        ///< of the _ebpf_release_epoch is complete and that future
+                                                        ///< epoch computation can now be initiated. In addition, it
+                                                        ///< signals the KEVENT associated with the message to signal
+                                                        ///< any waiting threads that the operation is completed.
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_EXIT_EPOCH, ///< This message is used when a thread running with IRQL < DISPATCH calls
+                                            ///< ebpf_epoch_exit on a different CPU than ebpf_epoch_enter. It is sent
+                                            ///< during from the thread that is existing the epoch to the CPU that
+                                            ///< ebpf_epoch_enter was called on.
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_RUNDOWN_IN_PROGRESS, ///< This message is sent to each CPU to notify it that epoch code
+                                                     ///< is shutting down and that no future timers should be armed and
+                                                     ///< future messages should be ignored.
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_IS_FREE_LIST_EMPTY,  ///< This message is sent to each CPU to query if it's local free
+                                                     ///< list is empty.
+} ebpf_epoch_cpu_message_type_t;
+
+/**
+ * @brief Message sent to a CPU to perform an action.
+ * Message can be one of the following:
```

Comment:
This does not seem to correlate to the enum above. For example, which message will cause the free-list to be emptied?

Reply:
Removed.